src/core/manager.c

```diff
@@ -1995,15 +1996,33 @@ int manager_load_unit_prepare(
 
         ret = manager_get_unit(m, name);
         if (ret) {
-                *_ret = ret;
-                return 1;
+                /* The time-based cache allows to start new units without daemon-reload,
+                 * but if they are already referenced (because of dependencies or ordering)
+                 * then we have to force a load of the fragment. As an optimization, check
+                 * first if anything in the usual paths was modified since the last time
+                 * the cache was loaded. */
+                if (ret->load_state == UNIT_NOT_FOUND && m->unit_cache_mtime > 0 &&
+                                !lookup_paths_mtime_good(&m->lookup_paths, m->unit_cache_mtime)) {
+                        ret->load_state = UNIT_STUB;
+                        r = unit_load_fragment(ret);
+                        if (r < 0 || ret->load_state == UNIT_STUB) {
+                                ret->load_state = UNIT_NOT_FOUND;
+                                *_ret = ret;
+                                return 1;
+                        }
+                } else {
+                        *_ret = ret;
+                        return 1;
+                }
+        } else {
+                ret = cleanup_ret = unit_new(m, unit_vtable[t]->object_size);
+                if (!ret)
+                        return -ENOMEM;
         }
 
-        ret = cleanup_ret = unit_new(m, unit_vtable[t]->object_size);
-        if (!ret)
-                return -ENOMEM;
-
         if (path) {
+                if (ret->fragment_path)
+                        free (ret->fragment_path);
```

Comment:
no space between function name and `(`

Reply:
more importantly though: free() is actually happy to accept a NULL argument, and is a NOP then. it's in ANSI C that way, hence the explicit if check is unnecessary.

We have free_and_strdup() as helper to combine the freeing and strdupping into one btw

Reply:
Ack, will change to free_and_strdup